fxa-auth-db-server/test/backend/db_tests.js

```diff
@@ -2325,6 +2326,79 @@ module.exports = function(config, DB) {
+            ]))
+            .then(results => {
+              results.forEach(r => t.deepEqual(r, {}, 'createSigninCode should return an empty object'))
+
+              // Attempt to create a duplicate code
+              return db.createSigninCode(SIGNIN_CODES[1], ACCOUNT.uid, TIMESTAMPS[1])
+                .then(() => t.fail('db.createSigninCode should fail for duplicate codes'))
+                .catch(err => {
+                  t.ok(err, 'db.createSigninCode should reject with an error')
+                  t.equal(err.code, 409, 'db.useSigninCode should reject with code 404')
+                  t.equal(err.errno, 101, 'db.useSigninCode should reject with errno 116')
+                })
+            })
+            .then(() => {
+              // Expire 2 of the codes
+              return db.expireSigninCodes(TIMESTAMPS[2])
+            })
+            .then(result => {
+              t.deepEqual(result, {}, 'expireSigninCodes should return an empty object')
+
+              // Attempt to use the 1st expired code
+              return db.useSigninCode(SIGNIN_CODES[0])
+                .then(() => t.fail('db.useSigninCode should fail for expired codes'))
+                .catch(err => {
+                  t.ok(err, 'db.useSigninCode should reject with an error')
+                  t.equal(err.code, 404, 'db.useSigninCode should reject with code 404')
+                  t.equal(err.errno, 116, 'db.useSigninCode should reject with errno 116')
+                })
+            })
+            .then(() => {
+              // Attempt to use the 2nd expired code
+              return db.useSigninCode(SIGNIN_CODES[1])
+                .then(() => t.fail('db.useSigninCode should fail for expired codes'))
+                .catch(err => {
+                  t.ok(err, 'db.useSigninCode should reject with an error')
+                  t.equal(err.code, 404, 'db.useSigninCode should reject with code 404')
+                  t.equal(err.errno, 116, 'db.useSigninCode should reject with errno 116')
+                })
+            })
+            .then(() => {
+              // Use the non-expired code
+              return db.useSigninCode(SIGNIN_CODES[2])
+            })
+            .then(result => {
+              t.deepEqual(result, { email: ACCOUNT.email }, 'db.useSigninCode should return an email address for non-expired codes')
+
+              // Attempt to re-use a used code
+              return db.useSigninCode(SIGNIN_CODES[2])
+                .then(() => t.fail('db.useSigninCode should fail for used codes'))
+                .catch(err => {
+                  t.ok(err, 'db.useSigninCode should reject with an error')
+                  t.equal(err.code, 404, 'db.useSigninCode should reject with code 404')
+                  t.equal(err.errno, 116, 'db.useSigninCode should reject with errno 116')
+                })
+            })
+            // Clean up the account
+            .then(() => db.deleteAccount(ACCOUNT.uid))
```

Comment:
Good point, that would have picked up my failure to point at the latest version of the `deleteAccount` stored procedure.